File: test/shire/schedules_test.exs
defmodule Shire.SchedulesTest do
  use Shire.DataCase, async: true
  use Oban.Testing, repo: Shire.Repo

  import Mox

  alias Shire.Schedules
  alias Shire.Schedules.ScheduledTask
  alias Shire.Agents
  alias Shire.Projects

  @vm Shire.VirtualMachineStub

  setup :set_mox_from_context

  setup do
    stub(Shire.VirtualMachineMock, :workspace_root, fn _project_id -> "/workspace" end)
    stub(Shire.VirtualMachineMock, :vm_status, fn _project_id -> :running end)

    {:ok, project} = Projects.create_project("schedule-project")
    {:ok, agent} = Agents.create_agent_with_vm(project.id, "sched-agent", "version: 1\n", @vm)

    %{project: project, agent: agent}
  end

  # Helper to insert a task directly, bypassing the Oban enqueue side-effect
  defp insert_task!(attrs) do
    %ScheduledTask{}
    |> ScheduledTask.changeset(attrs)
    |> Repo.insert!()
    |> Repo.preload(:agent)
  end

  describe "create_scheduled_task/1" do
    test "with valid recurring attrs", %{project: project, agent: agent} do
      attrs = %{
        label: "Daily check",
        message: "Run daily check",
        schedule_type: :recurring,
        cron_expression: "0 9 * * *",
        project_id: project.id,
        agent_id: agent.id
      }

      Oban.Testing.with_testing_mode(:manual, fn ->
        assert {:ok, task} = Schedules.create_scheduled_task(attrs)
        assert task.label == "Daily check"
        assert task.message == "Run daily check"
        assert task.schedule_type == :recurring
        assert task.cron_expression == "0 9 * * *"
        assert task.enabled == true
        assert task.project_id == project.id
        assert task.agent_id == agent.id
      end)
    end

    test "with valid once attrs", %{project: project, agent: agent} do
      scheduled_at = DateTime.utc_now() |> DateTime.add(3600) |> DateTime.truncate(:second)

      attrs = %{
        label: "One-time reminder",
        message: "Do this once",
        schedule_type: :once,
        scheduled_at: scheduled_at,
        project_id: project.id,
        agent_id: agent.id
      }

      Oban.Testing.with_testing_mode(:manual, fn ->
        assert {:ok, task} = Schedules.create_scheduled_task(attrs)
        assert task.label == "One-time reminder"
        assert task.schedule_type == :once
        assert task.scheduled_at == scheduled_at
      end)
    end

    test "with invalid cron expression", %{project: project, agent: agent} do
      attrs = %{
        label: "Bad cron",
        message: "Will fail",
        schedule_type: :recurring,
        cron_expression: "not a cron",
        project_id: project.id,
        agent_id: agent.id
      }

      assert {:error, changeset} = Schedules.create_scheduled_task(attrs)
      assert "is not a valid cron expression" in errors_on(changeset).cron_expression
    end

    test "with missing required fields" do
      assert {:error, changeset} = Schedules.create_scheduled_task(%{})
      errors = errors_on(changeset)
      assert "can't be blank" in errors.label
      assert "can't be blank" in errors.message
      assert "can't be blank" in errors.schedule_type
      assert "can't be blank" in errors.project_id
      assert "can't be blank" in errors.agent_id
    end

    test "recurring without cron_expression fails", %{project: project, agent: agent} do
      attrs = %{
        label: "Missing cron",
        message: "Should fail",
        schedule_type: :recurring,
        project_id: project.id,
        agent_id: agent.id
      }

      assert {:error, changeset} = Schedules.create_scheduled_task(attrs)
      assert "can't be blank" in errors_on(changeset).cron_expression
    end

    test "once without scheduled_at fails", %{project: project, agent: agent} do
      attrs = %{
        label: "Missing time",
        message: "Should fail",
        schedule_type: :once,
        project_id: project.id,
        agent_id: agent.id
      }

      assert {:error, changeset} = Schedules.create_scheduled_task(attrs)
      assert "can't be blank" in errors_on(changeset).scheduled_at
    end
  end

  describe "list_scheduled_tasks/1" do
    test "returns tasks for a project", %{project: project, agent: agent} do
      insert_task!(%{
        label: "Task A",
        message: "do A",
        schedule_type: :recurring,
        cron_expression: "0 * * * *",
        project_id: project.id,
        agent_id: agent.id
      })

      insert_task!(%{
        label: "Task B",
        message: "do B",
        schedule_type: :recurring,
        cron_expression: "30 * * * *",
        project_id: project.id,
        agent_id: agent.id
      })

      tasks = Schedules.list_scheduled_tasks(project.id)
      assert length(tasks) == 2
      labels = Enum.map(tasks, & &1.label)
      assert "Task A" in labels
      assert "Task B" in labels
    end

    test "does not return tasks from other projects", %{project: project, agent: agent} do
      {:ok, other_project} = Projects.create_project("other-project")

      insert_task!(%{
        label: "My task",
        message: "mine",
        schedule_type: :recurring,
        cron_expression: "0 * * * *",
        project_id: project.id,
        agent_id: agent.id
      })

      tasks = Schedules.list_scheduled_tasks(other_project.id)
      assert tasks == []
    end
  end

  describe "update_scheduled_task/2" do
    test "updates a task", %{project: project, agent: agent} do
      task =
        insert_task!(%{
          label: "Original",
          message: "original msg",
          schedule_type: :recurring,
          cron_expression: "0 * * * *",
          project_id: project.id,
          agent_id: agent.id
        })

      Oban.Testing.with_testing_mode(:manual, fn ->
        assert {:ok, updated} = Schedules.update_scheduled_task(task, %{label: "Updated"})
        assert updated.label == "Updated"
        assert updated.message == "original msg"
      end)
    end
  end

  describe "delete_scheduled_task/1" do
    test "deletes a task", %{project: project, agent: agent} do
      task =
        insert_task!(%{
          label: "Delete me",
          message: "bye",
          schedule_type: :recurring,
          cron_expression: "0 * * * *",
          project_id: project.id,
          agent_id: agent.id
        })

      assert {:ok, _} = Schedules.delete_scheduled_task(task)
      assert Schedules.get_scheduled_task(task.id) == nil
    end
  end

  describe "toggle_scheduled_task/2" do
    test "toggles enabled to false", %{project: project, agent: agent} do
      task =
        insert_task!(%{
          label: "Toggle me",
          message: "toggle",
          schedule_type: :recurring,
          cron_expression: "0 * * * *",
          project_id: project.id,
          agent_id: agent.id,
          enabled: true
        })

      Oban.Testing.with_testing_mode(:manual, fn ->
        assert {:ok, toggled} = Schedules.toggle_scheduled_task(task, false)
        assert toggled.enabled == false
      end)
    end

    test "toggles enabled to true", %{project: project, agent: agent} do
      task =
        insert_task!(%{
          label: "Toggle me",
          message: "toggle",
          schedule_type: :recurring,
          cron_expression: "0 * * * *",
          project_id: project.id,
          agent_id: agent.id,
          enabled: false
        })

      Oban.Testing.with_testing_mode(:manual, fn ->
        assert {:ok, toggled} = Schedules.toggle_scheduled_task(task, true)
        assert toggled.enabled == true
      end)
    end
  end

  describe "mark_run/1" do
    test "sets last_run_at", %{project: project, agent: agent} do
      task =
        insert_task!(%{
          label: "Run me",
          message: "run",
          schedule_type: :recurring,
          cron_expression: "0 * * * *",
          project_id: project.id,
          agent_id: agent.id
        })

      assert is_nil(task.last_run_at)
      assert {:ok, updated} = Schedules.mark_run(task)
      assert updated.last_run_at != nil
      assert updated.enabled == true
    end

    test "disables one-shot tasks", %{project: project, agent: agent} do
      scheduled_at = DateTime.utc_now() |> DateTime.add(3600) |> DateTime.truncate(:second)

      task =
        insert_task!(%{
          label: "Once only",
          message: "once",
          schedule_type: :once,
          scheduled_at: scheduled_at,
          project_id: project.id,
          agent_id: agent.id,
          enabled: true
        })

      assert {:ok, updated} = Schedules.mark_run(task)
      assert updated.last_run_at != nil
      assert updated.enabled == false
    end
  end

  describe "compute_next_run/1" do
    test "returns a future DateTime" do
      next_run = Schedules.compute_next_run("0 * * * *")
      assert %DateTime{} = next_run
      assert DateTime.compare(next_run, DateTime.utc_now()) == :gt
    end

    test "returns a DateTime for a specific cron" do
      next_run = Schedules.compute_next_run("30 9 * * 1")
      assert %DateTime{} = next_run
      assert DateTime.compare(next_run, DateTime.utc_now()) == :gt
    end
  end
end

File: src/services/schedules.test.ts
import { describe, it, expect, beforeEach } from "bun:test";
import { useTestDb } from "../test/setup";
import * as schedules from "./schedules";
import * as projects from "./projects";
import * as agents from "./agents";

describe("schedules service", () => {
  useTestDb();

  let projectId: string;
  let agentId: string;

  beforeEach(() => {
    const project = projects.createProject("schedule-project");
    projectId = project.id;
    const agent = agents.createAgent(projectId, { name: "sched-agent" });
    agentId = agent.id;
  });

  describe("createScheduledTask", () => {
    it("creates a recurring task", () => {
      const task = schedules.createScheduledTask({
        label: "Daily check",
        message: "Run daily check",
        scheduleType: "recurring",
        cronExpression: "0 9 * * *",
        projectId,
        agentId,
      });
      expect(task.label).toBe("Daily check");
      expect(task.message).toBe("Run daily check");
      expect(task.scheduleType).toBe("recurring");
      expect(task.cronExpression).toBe("0 9 * * *");
      expect(task.enabled).toBe(true);
    });

    it("creates a one-time task", () => {
      const scheduledAt = new Date(Date.now() + 3600_000).toISOString();
      const task = schedules.createScheduledTask({
        label: "One-time reminder",
        message: "Do this once",
        scheduleType: "once",
        scheduledAt,
        projectId,
        agentId,
      });
      expect(task.label).toBe("One-time reminder");
      expect(task.scheduleType).toBe("once");
      expect(task.scheduledAt).toBe(scheduledAt);
    });
  });

  describe("listScheduledTasks", () => {
    it("returns tasks for a project", () => {
      schedules.createScheduledTask({
        label: "Task A",
        message: "do A",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
      });
      schedules.createScheduledTask({
        label: "Task B",
        message: "do B",
        scheduleType: "recurring",
        cronExpression: "30 * * * *",
        projectId,
        agentId,
      });

      const tasks = schedules.listScheduledTasks(projectId);
      expect(tasks.length).toBe(2);
      const labels = tasks.map((t) => t.scheduled_tasks.label);
      expect(labels).toContain("Task A");
      expect(labels).toContain("Task B");
    });

    it("does not return tasks from other projects", () => {
      const otherProject = projects.createProject("other-project");
      schedules.createScheduledTask({
        label: "My task",
        message: "mine",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
      });

      const tasks = schedules.listScheduledTasks(otherProject.id);
      expect(tasks.length).toBe(0);
    });
  });

  describe("updateScheduledTask", () => {
    it("updates a task", () => {
      const task = schedules.createScheduledTask({
        label: "Original",
        message: "original msg",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
      });

      const updated = schedules.updateScheduledTask(task.id, { label: "Updated" });
      expect(updated?.label).toBe("Updated");
      expect(updated?.message).toBe("original msg");
    });
  });

  describe("deleteScheduledTask", () => {
    it("deletes a task", () => {
      const task = schedules.createScheduledTask({
        label: "Delete me",
        message: "bye",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
      });

      schedules.deleteScheduledTask(task.id);
      expect(schedules.getScheduledTask(task.id)).toBeUndefined();
    });
  });

  describe("toggleScheduledTask", () => {
    it("toggles enabled to false", () => {
      const task = schedules.createScheduledTask({
        label: "Toggle me",
        message: "toggle",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
        enabled: true,
      });

      const toggled = schedules.toggleScheduledTask(task.id, false);
      expect(toggled?.enabled).toBe(false);
    });

    it("toggles enabled to true", () => {
      const task = schedules.createScheduledTask({
        label: "Toggle me",
        message: "toggle",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
        enabled: false,
      });

      const toggled = schedules.toggleScheduledTask(task.id, true);
      expect(toggled?.enabled).toBe(true);
    });
  });

  describe("markRun", () => {
    it("sets last_run_at", () => {
      const task = schedules.createScheduledTask({
        label: "Run me",
        message: "run",
        scheduleType: "recurring",
        cronExpression: "0 * * * *",
        projectId,
        agentId,
      });

      expect(task.lastRunAt).toBeNull();
      const updated = schedules.markRun(task.id);
      expect(updated?.lastRunAt).toBeTruthy();
    });
  });
});
